dashboard/src/actions/overviewActions.js

```diff
@@ -5,12 +5,12 @@ import API from "../utils/axiosInstance";
 import { DANGER } from "assets/constants/toastConstants";
 import { findNoOfDays } from "utils/dateFunctions";
 import { showToast } from "./toastActions";
-import Cookies from "js-cookie";
 
 export const getDatasets = () => async (dispatch, getState) => {
   const alreadyRendered = getState().overview.loadingDone;
   try {
-    const username = Cookies.get("username");
+    const keycloak = getState().apiEndpoint.keycloak;
+    const username = keycloak?.idTokenParsed?.preferred_username;
```

Comment:
> You're probably going to get tired of me asking this, but is a `username` value of `undefined` acceptable, here? If not, don't be using

Yeah, I think it's acceptable if the keycloak server did not provide us with that value.

> If the `keycloak` object is part of the Redux state, then can't we count on it containing `idTokenParsed`? And, can't we count on that containing `preferred_username`? If not, then perhaps we should be making other arrangements, here. (Don't we really need the username for this code??...so how can we get by if it's `undefined`?)

If the user has not given permission to share their username with pbench-dashboard client then we can't do anything anyway. If they have given the permission already then the `idTokenParsed` will have it. So it doesn't matter whether we have `?` for the username or not.

Reply:
> Does the user have fine-grained control over what they share? I thought that the Keycloak client gets to say, "this is the list of things that you will share with me", and the user gets to say "yes" or "no". (And, if they say "no", I hope that the authorization fails!)

Again, it's up to how the Keycloak would be configured in the Intlab, let's not worry about it right now these are minor details. I believe even if the user doesn't have a username on the identity provider, the keycloak broker instance will use its email address as a username and it will provide us with some username. if the user denied sharing the basic data that the application client wants then yeah he/she will not be able to gain access to the dashboard's protected routes. In that case, yeah having a chaining operator will not make much difference as it will never return undefined so I guess can remove it.

Reply:
> let's not worry about it right now these are minor details

Knowing what information will or won't be available in the token is not a minor detail.  And, littering the code with conditional operations for cases which may or may not arise creates a testing and reliability problem.  For cases which aren't supposed to fail, it's much better to make the references unconditional and, if appropriate, include an error handler.  And, if we can know what information is going to be available, we can reduce the number of conditionals and error handlers required.  So, I really don't think it's a minor detail.

> it's up to how the Keycloak would be configured in the Intlab

Do you mean it's determined by how the Pbench Keycloak _client_ is configured on the Keycloak broker?  If so, that is under our (in)direct control, and we can stipulate what should be in the token.

> even if the user doesn't have a username on the identity provider, the keycloak broker instance will use its email address as a username and it will provide us with some username.

Good; then we don't need a conditional reference here (or elsewhere).

> if the user denied sharing the basic data that the application client wants then yeah he/she will not be able to gain access to the dashboard's protected routes.

I'm not quite connecting the dots:  if the user denies sharing the basic data that the client wants, do we get a redirect from the OIDC broker which says that the user is authenticated and provides us with a token?  (I.e., does `keycloak.authenticated` yield true or false?)

Reply:
> Do you mean it's determined by how the Pbench Keycloak _client_ is configured on the Keycloak broker? If so, that is under our (in)direct control, and we can stipulate what should be in the token.

The plan is that the intlab Keycloak is the broker we use, directly; it's a front end to Red Hat SSO. We own a Realm on the broker, but we don't control the broker instance aside from that.

However... yeah, stuff like this might be within our control under our Realm, but do we want to place hard external configuration requirements like that? I don't know ... it might violate someone's privacy rules, for example.

Reply:
I think we want to express the Pbench client's requirements, and then it is up to the user to decide whether to permit them.

Reply:
Oh wait a second, I know why I added chaining `?` operator in the `getDatasets` functions. It's because I don't want to fail on line 13 upon the overview page reload. On the reload of the overview page, the `keycloak` object here does not have the `idTokenParsed.preferred_username` attribute as it's not yet initialized. However, having the chaining operator will return an `undefined` string value for the owner which results in 401 from the axiosInstance, and therefore we get redirected back to the `home` page (and on the home page we get logged back in because of the check-sso). Failing on line 13 means we skip the `dataset_list` API call and remain on the overview page with an empty list of dataset without an option to refresh the datasets list.

Reply:
> I don't want to fail on line 13 upon the overview page reload. On the reload of the overview page, the keycloak object here does not have the idTokenParsed.preferred_username attribute as it's not yet initialized.

That's a good reason to use `?.`!  **_However_**, the code should then _test_ the value of `username` and do something appropriate if it is `undefined`.  It should _not_ just use it _as is_!